File: basic_dsa_three_hundred_seventy_seven.java
/**
 * Basic DSA Problems - Combination Sum IV (LeetCode 377)
 * Number of combinations that add up to target (order matters)
 */
public class basic_dsa_three_hundred_seventy_seven {

    // Problem: Distinct nums, target. Return count of sequences that sum to target.

    /**
     * DP: dp[i] = count of ways to make sum i. dp[i] += dp[i-n] for each n in nums.
     * Time O(target * n), Space O(target)
     */
    public static int combinationSum4(int[] nums, int target) {
        int[] dp = new int[target + 1];
        dp[0] = 1;
        for (int i = 1; i <= target; i++)
            for (int n : nums)
                if (i >= n) dp[i] += dp[i - n];
        return dp[target];
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Combination Sum IV");
        System.out.println("=".repeat(60));

        System.out.println("nums=[1,2,3], target=4 -> " + combinationSum4(new int[]{1,2,3}, 4));
        System.out.println("nums=[9], target=3 -> " + combinationSum4(new int[]{9}, 3));
    }
}

File: basic_dsa_three_hundred_seventy_seven.cpp
/**
 * Basic DSA - Sum of Square Numbers (LeetCode 633)
 * Return true if c == a^2 + b^2 for integers a, b >= 0.
 */
#include <cmath>
#include <iostream>
#include <string>

bool judgeSquareSum(int c) {
    long long lo = 0;
    long long hi = static_cast<long long>(std::sqrt(c));
    while (lo <= hi) {
        long long s = lo * lo + hi * hi;
        if (s == c)
            return true;
        if (s < c)
            ++lo;
        else
            --hi;
    }
    return false;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Sum of Square Numbers\n";
    std::cout << std::string(60, '=') << "\n";

    std::cout << "c=5 -> " << (judgeSquareSum(5) ? "true" : "false") << "\n";
    std::cout << "c=3 -> " << (judgeSquareSum(3) ? "true" : "false") << "\n";
    std::cout << "c=0 -> " << (judgeSquareSum(0) ? "true" : "false") << "\n";
}
